primitives/state-machine/src/ext.rs

```diff
@@ -792,31 +793,72 @@ impl Encode for EncodeOpaqueValue {
 	}
 }
 
+// AppendValue
+#[derive(Debug,PartialEq)]
+struct AppendedValue {
+	pub num: Compact<u32>,
+	pub data: Vec<u8>,
+	pub materialized: Option<Vec<u8>>
+}
+
+
+
+impl Default for AppendedValue {
+	fn default() -> Self{
+		AppendedValue {
+			num: Compact(0),
+			data: vec![],
+			materialized: None
+		}
+	}
+}
+
 /// Auxialiary structure for appending a value to a storage item.
-pub(crate) struct StorageAppend<'a>(&'a mut Vec<u8>);
+pub(crate) struct StorageAppend(AppendedValue);
 
-impl<'a> StorageAppend<'a> {
+
+
+impl StorageAppend {
 	/// Create a new instance using the given `storage` reference.
-	pub fn new(storage: &'a mut Vec<u8>) -> Self {
-		Self(storage)
+	pub fn new(storage: &mut Vec<u8>) -> Self {
+		let appended_value = AppendedValue {
+			num:Compact(0),
+			data: storage.to_vec(),
+			materialized: None
+		};
+		Self(appended_value)
 	}
 
-	/// Append the given `value` to the storage item.
-	///
-	/// If appending fails, `[value]` is stored in the storage item.
+	// Append the given `value` to the storage item.
+	//
+	// If appending fails, `[value]` is stored in the storage item.
 	pub fn append(&mut self, value: Vec<u8>) {
 		let value = vec![EncodeOpaqueValue(value)];
 
-		let item = sp_std::mem::take(self.0);
+		let item = sp_std::mem::take(&mut self.0);
 
-		*self.0 = match Vec::<EncodeOpaqueValue>::append_or_new(item, &value) {
-			Ok(item) => item,
+		self.0 = match Vec::<EncodeOpaqueValue>::append_or_new(item.data, &value) {
+			Ok(data) => {
+				let num = <u32>::from(item.num).saturating_add(1);
+				let num = Compact::<u32>::from(num);
+				let appended_value = AppendedValue{
+					num,
+					data,
+					materialized: None
+				};
+				appended_value
+			},
```

Comment:
The code you have added here doesn't make that much sense. You should not call `append_or_new` anymore. You want to move the materialization close to the point where the runtime actually calls get on this data or we want to store it in the state.

In general you should make yourself familiar with `OverlayedChanges` and how that works. Especially in the light of storage transaction. Currently both of these things just work with `Vec<u8>`s. This would need to be changed to have values either be some `Vec<u8>` or a `AppendedValue`. 

The entire point of the linked issue is to improve the performance of the `append` operation, especially when it comes to storage transactions.

Reply:
Okey and so I need to make that if runtime request data it should first check materialized because it will contain latest data and if not it will go on checking data field. 
And when storing I should first update materialized  field then go on appending the data into the `vec`.
And on `OverlayedChanges ` shouldn't be only working with `AppendValue` because inside it it will have data filed with `vec<u8>` ?

Reply:
And am into it will make some changes and iterate

Reply:
> And on `OverlayedChanges ` shouldn't be only working with `AppendValue` because inside it it will have data filed with `vec<u8>` ?

I don't get this question.

> And when storing I should first update materialized field then go on appending the data into the `vec`.

Materialized should be set to `None` when appending new data.

> Okey and so I need to make that if runtime request data it should first check materialized because it will contain latest data and if not it will go on checking data field.

First check materialized and if that doesn't exist, create materialized from the data + length.

Reply:
Check the new commit, does not compile yet , But i wanted to make sure that am heading straight. 

So at first `OverlayedChanges` only worked with `vec<u8>` but Ive changed to work  with `AppendedValue` but also I will have to change numerous APIs.  So while appending it increments the `num` filled and changes the `data filled`